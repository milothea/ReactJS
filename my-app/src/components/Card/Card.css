.container.card {
    width: 15rem;
    height: 10rem;
    padding: 1rem;
    margin: .5rem;
    text-align: left;
    border: .15rem solid #f7e773;
    border-radius: .5rem;
    background-color: #ffffff;
    box-shadow: 0 0 .2rem rgba(94, 94, 94, .2);
}

.disable-mode .container.card { background-color: #fbf7d8; }

.container.card.active {
    background-color: #333;
    color: #ffffff;
}

.container.card.editmode { background-color: #eeedeb; }
