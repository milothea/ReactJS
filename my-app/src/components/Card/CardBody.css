.card__text {
    display: block;
    width: 14rem;
    height: 7.5rem;
    font-size: .8rem;
    word-wrap: break-word;
    padding: .5rem;
    overflow: auto;
}

.editmode .card__text {
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
    padding: .4rem;
    background-color: #f9f8f7;
    border: .1rem solid transparent;
    border-radius: .15rem;
    color: #656565;
    resize: none;
}

.editmode .card__text {
    width: 14rem;
    height: 7rem;
    padding-top: .2rem;
    margin-top: .2rem;
}
