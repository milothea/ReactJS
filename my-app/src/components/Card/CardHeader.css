.disable-mode .card__heading { width: 13.2rem; }

.container.heading {
    width: 100%;
    height: 2.5rem;
    display: flex;
    flex-wrap: nowrap;
    padding: 0;
    border-bottom: .063rem solid #f7e773;
}

.card__heading {
    padding: .5rem;
    text-transform: capitalize;
}

.card__heading {
    display: block;
    width: 12rem;
    height: auto;
    font-size: 1.1rem;
    line-height: 1.1rem;
    word-wrap: break-word;
    overflow: auto;
}

.editmode .card__heading {
    height: 1rem;
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
    padding: .45rem;
    background-color: #f9f8f7;
    border: .1rem solid transparent;
    border-radius: .15rem;
    color: #656565;
    font-weight: 700;
    text-transform: none;
}

.card__checkbox { margin-top: .6rem; }

.card__checkbox:hover { cursor: pointer; }

.card__edit,
.card__submit,
.card__cancel { height: 2.1rem; }

.card__edit:hover,
.card__submit:hover,
.card__cancel:hover { cursor: pointer; }

.card__edit,
.card__submit { padding-right: .5rem; }

.card__edit {
    color: #656565;
    fill: #f7e773;
}

.container.card.active .card__edit { color: #999999; }

.card__edit:hover { fill: #e0a7f7; }

/* Styles for correct view in Mozilla Firefox */
@-moz-document url-prefix() {
    .card__heading { line-height: .95rem; }

    .card__edit { height: 2.45rem; }

    .card__checkbox { margin-top: 0; }
}
