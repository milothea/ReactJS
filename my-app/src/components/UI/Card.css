.container__card {
    width: 15rem;
    height: 10rem;
    padding: 1rem;
    text-align: left;
    border: .15rem solid #f7e773;
    border-radius: .5rem;
    background-color: #ffffff;
    box-shadow: 0 0 .2rem rgba(94, 94, 94, .2);
}

.card__heading,
.card__text {
    padding: .5rem;
    text-transform: capitalize;
}

.card__heading {
    border-bottom: .063rem solid #f7e773;
    line-height: 2rem;
}
