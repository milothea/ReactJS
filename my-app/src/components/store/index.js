import { createStore } from 'redux';
import { v4 as uuidv4 } from 'uuid';

const defaultState = {
    appPaths: {
        mainPage: '/',
        errorPage: '/error',
        authPage: '/authorisation'
    },
    cardsData: [],
    isLoggedIn: false
};

const appReducer = (state = defaultState, action) => {
    switch (action.type) {
        case 'cardsData/setData':
            state.cardsData = action.payload.data;
            return state;
        case 'cardsData/updateCardData':
            state.cardsData = state.cardsData.map(card => {
               if (card.Number === action.payload.id) {
                   const data = {...card};

                   data.Name = action.payload.newHeading;
                   data.About = action.payload.newText;

                   return data;
               }

               return card;
            });

            return state;
        case 'cardsData/switchActiveState':
            state.cardsData = state.cardsData.map(card => {
                if (card.Number === action.payload.id) {
                    const prev = {...card};

                    prev.isActive = !card.isActive;

                    return prev;
                }

                return card;
            });

            return state;
        case 'cardsData/deleteCards':
            state.cardsData = state.cardsData.filter(card => !(card.isActive));

            return state;
        case 'cardsData/addCard':
            state.cardsData = [...state.cardsData, {
                Name: action.payload.heading,
                About: action.payload.text,
                isActive: false,
                Number: uuidv4()
            }];

            return state;
        case 'authorization/login':
            state.isLoggedIn = true;

            return state;
        default:
            return state;
    }
};

const store = createStore(appReducer);

export default store;
