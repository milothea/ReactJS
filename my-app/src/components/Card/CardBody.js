import './CardBody.css';

const CardBody = ({ value, isEditMode, onChange }) => {

    return isEditMode ? (
        <textarea className='card__text'
                  value={value}
                  maxLength='140'
                  onChange={onChange}
                  data-testid='textarea' />
    ) : (
        <p className='card__text'
           data-testid='paragraph'>{value}</p>
    );
};

export default CardBody;
