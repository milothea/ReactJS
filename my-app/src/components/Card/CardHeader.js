import './CardHeader.css';
import { HiOutlineCheck, HiOutlinePencil, HiOutlineX} from "react-icons/hi";

const CardHeader = ({
                        isActive,
                        isEditMode,
                        isReadOnly,
                        value,
                        onEdit,
                        onChange,
                        onCancel,
                        onSubmit,
                        onChecked
                    }) => {
    return isEditMode ? (
        <div className='container heading'>
            <input className='card__heading'
                   defaultValue={value}
                   maxLength='15'
                   onChange={onChange}
                   data-testid='input' />
            <HiOutlineCheck className='card__submit'
                            onClick={onSubmit}
                            data-testid='submit-icon' />
            <HiOutlineX className='card__cancel'
                        onClick={onCancel}
                        data-testid='cancel-icon' />
        </div>
    ) : (
        <div className='container heading'>
            <h2 className='card__heading'>{value}</h2>
            { !isReadOnly ? (
                <HiOutlinePencil className='card__edit'
                                 fill='red'
                                 onClick={onEdit}
                                 data-testid='edit-icon'/>
                ) : ''
            }
            <input className='card__checkbox'
                   type='checkbox'
                   defaultChecked={isActive}
                   onChange={onChecked}
                   data-testid='checkbox' />
        </div>
    );
}

export default CardHeader;
