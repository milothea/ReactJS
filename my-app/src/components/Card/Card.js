import React, { useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';
import { cardsActions } from '../store/cardsDataSlice';
import PropTypes from 'prop-types';
import './Card.css';
import CardHeader from './CardHeader';
import CardBody from './CardBody';

const Card = ({
                  id,
                  isReadOnly,
                  isActive,
                  heading,
                  text
}) => {
    const dispatch = useDispatch();
    const [isEditMode, setIsEditMode] = useState(false);
    const [headingInputValue, setHeadingInputValue] = useState(heading);
    const [textInputValue, setTextInputValue] = useState(text);

    const editHandler = () => {
        setIsEditMode(true);
        if (isActive) dispatch(cardsActions.toggleActiveState({ id: id }));
    };

    const headingInputHandler = (event) => setHeadingInputValue(event.target.value);

    const textInputHandler = (event) => setTextInputValue(event.target.value);

    const cancelHandler = () => {
        setIsEditMode(false);
        setHeadingInputValue(heading);
        setTextInputValue(text);
    };

    const submitHandler = () => {
        setIsEditMode(false);
        dispatch(cardsActions.updateCardData({
            id: id,
            newHeading: headingInputValue,
            newText: textInputValue
        }));
    };

    const changeCheckboxHandler = () => dispatch(cardsActions.toggleActiveState({ id: id}));

    useEffect(() => setIsEditMode(false), [isReadOnly]);

    return (
        <div className={`container card${isActive ? ' active' : ''}${isEditMode ? ' editmode' : ''}`}
             id={id}>
            <CardHeader id={id}
                        value={isEditMode ? headingInputValue : heading}
                        isReadOnly={isReadOnly}
                        isActive={isActive}
                        isEditMode={isEditMode}
                        onEdit={editHandler}
                        onChange={headingInputHandler}
                        onCancel={cancelHandler}
                        onSubmit={submitHandler}
                        onChecked={changeCheckboxHandler}
            />
            <CardBody isEditMode={isEditMode}
                      value={isEditMode ? textInputValue : text}
                      onChange={textInputHandler}
            />
        </div>
    );
}

Card.propTypes = {
    id: PropTypes.string,
    isReadOnly: PropTypes.bool,
    isActive: PropTypes.bool,
    heading: PropTypes.string,
    text: PropTypes.string,
    onUpdateCardData: PropTypes.func,
    onChangeActiveState: PropTypes.func
};

export default React.memo(Card);
