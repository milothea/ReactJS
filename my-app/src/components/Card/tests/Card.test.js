import * as react from 'react';
import * as redux from 'react-redux';
import Enzyme, {mount, shallow} from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import store from '../../store';
import Card from '../Card';
import CardHeader from '../CardHeader';
import CardBody from '../CardBody';

Enzyme.configure({ adapter: new Adapter() });

describe('Card component', () => {
    const testHeading = 'some heading';
    const testText = 'some text';
    const testId = '10';

    test('renders correctly in NOT read only mode and NOT active mode', () => {
       const component = renderer.create(
           <Provider store={store}>
               <Card id={testId}
                     isReadOnly={false}
                     isActive={false}
                     heading={testHeading}
                     text={testText} />
           </Provider>
       ).toJSON();

       expect(component).toMatchSnapshot();
    });

    test('renders correctly in read only mode and NOT active mode', () => {
        const component = renderer.create(
            <Provider store={store}>
                <Card id={testId}
                      isReadOnly={true}
                      isActive={false}
                      heading={testHeading}
                      text={testText} />
            </Provider>
        ).toJSON();

        expect(component).toMatchSnapshot();
    });

    test('renders correctly in read only mode and active mode', () => {
        const component = renderer.create(
            <Provider store={store}>
                <Card id={testId}
                      isReadOnly={true}
                      isActive={true}
                      heading={testHeading}
                      text={testText} />
            </Provider>
        ).toJSON();

        expect(component).toMatchSnapshot();
    });
});

describe('Card methods: ', () => {
    const id = 'test-id';
    let setIsEditMode;
    let setHeadingInputValue;
    let setTextInputValue;
    let mockedDispatch;
    let component;
    let cardHeader;

    beforeEach(() => {
        setIsEditMode = jest.fn();
        jest.spyOn(react, 'useState').mockImplementation(setIsEditMode);

        mockedDispatch = jest.fn();
        redux.useDispatch = () => mockedDispatch;

        component = mount(
            <Card id={id}
                  isReadOnly={false}
                  isActive={false} />
        );
        cardHeader = component.find(CardHeader);
    });

    test('editHandler works', () => {
        const editFunc = cardHeader.props().onEdit;

        expect(typeof editFunc).toBe('function');
    });

    test('cancelHandler works', () => {
        const cancelFunc = cardHeader.props().onCancel;

        expect(typeof cancelFunc).toBe('function');
        // expect(setIsEditMode).toBeCalledWith(false);
        // expect(setHeadingInputValue).toBeCalledTimes(1);
        // expect(setTextInputValue).toBeCalledTimes(1);
    });

    test('submitHandler works', () => {
        const submitFunc = cardHeader.props().onSubmit;

        submitFunc();
        expect(typeof submitFunc).toBe('function');
        // expect(setIsEditMode).toBeCalledTimes(1);
        expect(mockedDispatch).toBeCalledTimes(1);
    });

    test('changeCheckboxHandler works', () => {
        const checkFunc = cardHeader.props().onChecked;

        checkFunc();
        expect(typeof checkFunc).toBe('function');
        expect(mockedDispatch).toBeCalledTimes(1);
        expect(mockedDispatch.mock.calls[0][0].payload).toEqual({id: id});
    });

    test('textInputHandler is a function', () => {
       const inputHandler = component.find(CardBody).props().onChange;

       expect(typeof inputHandler).toBe('function');
    });
});
