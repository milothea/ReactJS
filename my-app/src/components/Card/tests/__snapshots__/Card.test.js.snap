// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component renders correctly in NOT read only mode and NOT active mode 1`] = `
<div
  className="container card"
  id="10"
>
  <div
    className="container heading"
  >
    <h2
      className="card__heading"
    >
      some heading
    </h2>
    <svg
      className="card__edit"
      data-testid="edit-icon"
      fill="red"
      height="1em"
      onClick={[Function]}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M15.232 5.232l3.536 3.536m-2.036-5.036a2.5 2.5 0 113.536 3.536L6.5 21.036H3v-3.572L16.732 3.732z"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
      />
    </svg>
    <input
      className="card__checkbox"
      data-testid="checkbox"
      defaultChecked={false}
      onChange={[Function]}
      type="checkbox"
    />
  </div>
  <p
    className="card__text"
    data-testid="paragraph"
  >
    some text
  </p>
</div>
`;

exports[`Card component renders correctly in read only mode and NOT active mode 1`] = `
<div
  className="container card"
  id="10"
>
  <div
    className="container heading"
  >
    <h2
      className="card__heading"
    >
      some heading
    </h2>
    
    <input
      className="card__checkbox"
      data-testid="checkbox"
      defaultChecked={false}
      onChange={[Function]}
      type="checkbox"
    />
  </div>
  <p
    className="card__text"
    data-testid="paragraph"
  >
    some text
  </p>
</div>
`;

exports[`Card component renders correctly in read only mode and active mode 1`] = `
<div
  className="container card active"
  id="10"
>
  <div
    className="container heading"
  >
    <h2
      className="card__heading"
    >
      some heading
    </h2>
    
    <input
      className="card__checkbox"
      data-testid="checkbox"
      defaultChecked={true}
      onChange={[Function]}
      type="checkbox"
    />
  </div>
  <p
    className="card__text"
    data-testid="paragraph"
  >
    some text
  </p>
</div>
`;
